src/hotspot/share/compiler/compileBroker.cpp

```diff
@@ -1355,6 +1355,9 @@ nmethod* CompileBroker::compile_method(const methodHandle& method, int osr_bci,
     method->constants()->resolve_string_constants(CHECK_AND_CLEAR_NONASYNC_NULL);
     // Resolve all classes seen in the signature of the method
     // we are compiling.
+    // This might result in loading a class that we're already loading resulting
+    // in a ClassCircularityError which is why the class loading code detects
+    // this in the loading path and not just the load_super_or_fail path.
     Method::load_signature_classes(method, CHECK_AND_CLEAR_NONASYNC_NULL);
```

Comment:
FTR I'm experimenting with removing eager loading of signature classes.

Reply:
Oh interesting, it might help. I don't know if there are other -Xcomp paths that start loading a class while it's already loading that class.
Sorry I didn't see  your comments since this is a draft and all the notifications look like bot notifications so I filter them.  Thanks for looking at this.